AzureMonitorAgent/agent.py

```diff
@@ -668,16 +680,122 @@ def metrics():
 
     return 0, ""
 
+
+def start_arc_process():
+    """
+    Start arc process that performs periodic monitoring activities
+    :return: None
+
+    """
+    hutil_log_info("stopping previously running arc process")
+    stop_arc_watcher()
+    hutil_log_info("starting arc process")
+    
+    #start telemetry watcher
+    oneagent_filepath = os.path.join(os.getcwd(),'agent.py')
+    args = ['python', oneagent_filepath, '-arc']
+    log = open(os.path.join(os.getcwd(), 'daemon.log'), 'w')
+    HUtilObject.log('start watcher process '+str(args))
+    subprocess.Popen(args, stdout=log, stderr=log)
+
+def start_arc_watcher():
+    """
+    Take care of starting arc_watcher daemon if the VM has arc running
+    """    
+    hutil_log_info("Starting the watcher")
+    print("Starting the watcher")
+    pids_filepath = os.path.join(os.getcwd(), 'amaarc.pid')
+    py_pid = os.getpid()
+    print("pid ", py_pid)
+    with open(pids_filepath, 'w') as f:
+        f.write(str(py_pid) + '\n')
+    hutil_log_info("Written all the pids")
+    print("Written all the pids")
+    watcher_thread = Thread(target = arc_watcher, args = [HUtilObject.error, HUtilObject.log])
+    watcher_thread.start()
+    watcher_thread.join()
+
+    return 0, ""
+
 # Dictionary of operations strings to methods
 operations = {'Disable' : disable,
               'Uninstall' : uninstall,
               'Install' : install,
               'Enable' : enable,
               'Update' : update,
               'Metrics' : metrics,
+              'Arc' : start_arc_watcher,
 }
 
 
+def stop_arc_watcher():
+    """
+    Take care of stopping arc_watcher daemon if the VM has arc running
+    """    
+    pids_filepath = os.path.join(os.getcwd(),'amaarc.pid')
+
+    # kill existing telemetry watcher
+    if os.path.exists(pids_filepath):
+        with open(pids_filepath, "r") as f:
+            for pids in f.readlines():
+                proc = subprocess.Popen(["ps -o cmd= {0}".format(pids)], stdout=subprocess.PIPE, shell=True)
+                output = proc.communicate()[0]
+                if "arc" in output:
+                    kill_cmd = "kill " + pids 
+                    run_command_and_log(kill_cmd)
+        
+        # Delete the file after to avoid clutter
+        os.remove(pids_filepath)
+
+def arc_watcher(hutil_error, hutil_log):
+    """
+    This is needed to override mdsd's syslog permissions restriction which prevents mdsd 
+    from reading temporary key files that are needed to make https calls to get an MSI token for arc during onboarding to download amcs config
+    This method spins up a process that will continuously keep refreshing that particular file path with valid keys
+    So that whenever mdsd needs to refresh it's MSI token, it is able to find correct keys there to make the https calls
+    """
+    # check every 25 seconds
+    sleepTime =  25
+
+    # sleep before starting the monitoring.
```

Comment:
verify this

Reply:
Is this for that whole verification on RHEL or just this specific code block?

Reply:
both :)

Reply:
Tested both on RHEL. Works as expected